Add documentation for ransomware (#1265)
guardicore/monkey: docs/content/reference/ransomware.md

---
title: "Ransomware"
date: 2021-06-23T18:13:59+05:30
draft: true
pre: '<i class="fas fa-lock"></i> '
weight: 10
---

The Infection Monkey has the capability of simulating a ransomware attack on your network.

> All actions performed by the encryption routine are safe for production environments.

Reviewer: Maybe we soften this a bit. This is a subtle change, but I think it matters.

```suggestion
All actions performed by the encryption routine are designed to be safe for production environments.
```

> To ensure minimum intereference and easy recoverability, the ransomware simulation will only run if
> it is configured properly. To do so, you must specify the path to a directory in the configuration.
> If no directory is specified, the simulation will not run.

Reviewer: The simulation is more than just encryption, it is also propagation. It potentially includes changing wallpaper or enumerating files.
```suggestion
To ensure minimum intereference and easy recoverability, the ransomware simulation will encrypt
files if it the user specifies a directory that contains files that are safe to encrypt.
If no directory is specified, no files will be encrypted.
```

Reviewer 2: I think that we also need to add that the user can't choose certain directories on his system.

Reviewer: ATM, we're not planning on adding any such restriction.

<!-- add config screenshot here -->

## Which files are encrypted?

All regular files with [relevant extensions](#relevant-file-extensions-for-encryption) in the
configured directory are attempted to be encrypted during the simulation.

The simulation is not recursive, i.e. it will not touch any files in sub-directories of the
configured directory — only appropriate files in the top level of the tree.

Symlinks and shortcuts are ignored.

## How are the files encrypted?

Files are "encrypted" in place with a simple bit flip. Encrypted files are renamed to have

> `.m0nk3y` appended to their names.

Reviewer 2: The `.m0nk3y` extension is the default one right? Shouldn't we also say that that can be changed ?

Reviewer: We haven't added the capability to choose your own yet. It might not make it into this release.

Author: It's best to update documentation when (or shortly after) functionality changes are made instead of before.

> To "decrypt" your files, you can simply perform a bit flip on them again.

Reviewer: I think this information needs to be moved higher up, and more information needs to be given about how this is "safe" and why it's sufficient for a ransomware simulation.

## Relevant file extensions for encryption

Encryption attempts are only performed on regular files with the following extensions.

> This list is based on the [analysis of the ??? ransomware by ???]().

Reviewer: ... analysis of the Goldeneye ransomware by BitDefender_

https://labs.bitdefender.com/2017/07/a-technical-look-into-the-goldeneye-ransomware-attack/

---
title: "Ransomware"
date: 2021-06-23T18:13:59+05:30
draft: true
pre: '<i class="fas fa-lock"></i> '
weight: 10
---

The Infection Monkey has the capability of simulating a ransomware attack on your network.
All actions performed by the encryption routine are designed to be safe for production
environments.

To ensure minimum interference and easy recoverability, the ransomware simulation will encrypt
files only if the user specifies a directory that contains files that are safe to encrypt.
If no directory is specified, no files will be encrypted.

<!-- add config screenshot here -->

## How are the files encrypted?

Files are "encrypted" in place with a simple bit flip. Encrypted files are renamed to have
`.m0nk3y` appended to their names.

This is a safe way to simulate encryption since it is easy to "decrypt" your files. You can simply perform a bit flip on the files again and rename them to remove the appended `.m0nk3y` extension.

> This is sufficient to mock a ransomware attack on your network as your files are left unusuable and are renamed with a different extension, similar to the way that many ransomwares act. As this is a simulation, your security solutions should be triggered to notify and prevent these changes from taking place.

Reviewer 2: ```suggestion
This is sufficient to mock a ransomware attack on your network as your files are left "unusable" and are renamed with a different extension, similar to the way that many ransomwares act. As this is a simulation, your security solutions should be triggered to notify and prevent these changes from taking place.
```

Reviewer 2: The users need to be sure that the files are not actually unusable.

Author: I think the paragraph right before this one specifies that well enough?

Author: I think the paragraph right before this one specifies that clearly enough. WDYT?

Reviewer: What if we say "the data in the files has been manipulated" instead of "your files are left unusable".

Reviewer 2: > I think the paragraph right before this one specifies that clearly enough. WDYT?

Yes it does. I just said so the user reading this will not be scared.

I like what @mssalvatore suggested.

Author: How about now?
https://github.com/guardicore/monkey/pull/1265/commits/954cc469cf7cab9c50c18b312fa9741895abd235

Reviewer: I'm good with this change. I'm going to have marketing review this before it gets released as well, so I don't think we have to agonize over every word :)

---
title: "Ransomware"
date: 2021-06-23T18:13:59+05:30
draft: true
pre: '<i class="fas fa-lock"></i> '
weight: 10
---

The Infection Monkey has the capability of simulating a ransomware attack on your network.
All actions performed by the encryption routine are designed to be safe for production
environments.

To ensure minimum interference and easy recoverability, the ransomware simulation will encrypt

> files if the user specifies a directory that contains files that are safe to encrypt.

Reviewer: ```suggestion
files only if the user specifies a directory that contains files that are safe to encrypt.
```

> If no directory is specified, no files will be encrypted.

Reviewer 3: @mssalvatore is this important enough to be added to the description of directory input fields in config UI?!

Author: I feel it is. What if the user hasn't read the documentation?

Reviewer: Within reason, I think a self-documenting UI is better than just relying on just documentation. I vote yes on adding this information to the UI. How do we do this now, a "description" field?

Author: Yes, the same way we're doing it for custom PBA commands.

Reviewer: @VakarisZ @shreyamalviya I've added a task to #1239 to make sure we don't forget.

> <!-- add config screenshot here -->

Reviewer 3: Add it?

Author: We haven't implemented it yet, hence, the comment. :)

## How are the files encrypted?

Files are "encrypted" in place with a simple bit flip. Encrypted files are renamed to have
`.m0nk3y` appended to their names.

This is a safe way to simulate encryption since it is easy to "decrypt" your files. You can simply perform a bit flip on the files again and rename them to remove the appended `.m0nk3y` extension.

This is sufficient for a ransomware simulation as your files are unusuable and are renamed with a different extension, similar to how many ransomwares act. These changes should trigger your security solutions.

## Which files are encrypted?

All regular files with [relevant extensions](#relevant-file-extensions-for-encryption) in the
configured directory are attempted to be encrypted during the simulation.

> The simulation is not recursive, i.e. it will not touch any files in sub-directories of the

Reviewer: Maybe just another note about how this prevents monkey from going rogue and encrypting files you didn't intend it to.

> configured directory — only appropriate files in the top level of the tree.

Reviewer: "top level of the tree" implies `C:\` or `/` to me, which would scare users and isn't what we mean here.

Symlinks and shortcuts are ignored.

> ## Relevant file extensions for encryption

Reviewer: Maybe "allowed" instead of "relevant"

Reviewer 3: Not "allowed", because the user doesn't need to allow it, they will simply get encrypted if found. So maybe "File extensions targeted" or something.

Encryption attempts are only performed on regular files with the following extensions.

This list is based on the [analysis of the Goldeneye ransomware by BitDefender](https://labs.bitdefender.com/2017/07/a-technical-look-into-the-goldeneye-ransomware-attack/).

- .3ds
- .7z
- .accdb
- .ai
- .asp
- .aspx
- .avhd
- .avi
- .back
- .bak
- .c
- .cfg
- .conf
- .cpp
- .cs
- .ctl
- .dbf
- .disk
- .djvu
- .doc
- .docx
- .dwg
- .eml
- .fdb
- .giff
- .gz
- .h
- .hdd
- .jpg
- .jpeg
- .kdbx
- .mail
- .mdb
- .mpg
- .mpeg
- .msg
- .nrg
- .ora
- .ost
- .ova
- .ovf
- .pdf
- .php
- .pmf
- .png
- .ppt
- .pptx
- .pst
- .pvi
- .py
- .pyc
- .rar
- .rtf
- .sln
- .sql
- .tar
- .tiff
- .txt
- .vbox
- .vbs
- .vcb
- .vdi
- .vfd
- .vmc
- .vmdk
- .vmsd
- .vmx
- .vsdx
- .vsv
- .work
- .xls
- .xlsx
- .xvd

> - .zip

Reviewer 3: @mssalvatore We definitely want to link here from config UI, otherwise the user will add files to the directory, they won't get encrypted and we'll get support emails about our software not working. Not all users go to the documentation and read everything they can find about a single configuration field. More likely than not, they don't even properly read descriptions...

Reviewer: I've added a task to #1239
